src/Microsoft.ML.Data/Training/TrainerInputBase.cs

```diff
@@ -0,0 +1,197 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using Microsoft.Data.DataView;
+using Microsoft.ML.Calibrators;
+using Microsoft.ML.CommandLine;
+using Microsoft.ML.Data;
+using Microsoft.ML.Data.IO;
+using Microsoft.ML.EntryPoints;
+
+namespace Microsoft.ML.Trainers
+{
+    /// <summary>
+    /// The base class for all trainer inputs.
+    /// </summary>
+    [TlcModule.EntryPointKind(typeof(CommonInputs.ITrainerInput))]
+    public abstract class TrainerInputBase
```

Comment:
>TrainerInputBase [](start = 26, length = 16)

This class should have a `[BestFriend] private protected` constructor. We do not want people forming base classes from it willy nilly. Similar with all the other abstract classes defined here.

Reply:
This could potentially be part of another PR. But, it's relatively easy to do if you have to change the PR for some other reason anyway.

---
In reply to: [260837670](https://github.com/dotnet/machinelearning/pull/2748#discussion_r260837670) [](ancestors = 260837670)